### ISSUE TITLE
Harvard-Oxford atlas: some atlases have wrong number of labels

### ISSUE THREAD

**Issue by illdopejake** [nilearn/nilearn#2036]
I'm was working on harmonizing the atlas outputs, but I found that some (5/16) of the Harvard-Oxford maps have the wrong number of labels. In other words, the number of atlas.labels does not match the number of np.unique(atlas.maps.get_data()). 

```python
In [80]: for name in names: 
    ...:     x = datasets.atlas.fetch_atlas_harvard_oxford(name) 
    ...:     img = image.load_img(x.maps).get_data() 
    ...:     try: 
    ...:         assert(all(np.unique(img) == list(range(len(np.unique(img)))))) 
    ...:     except: 
    ...:         print('{}: atlas labels are not a susquential list of' 
    ...:               ' numbers starting at 0'.format(name)) 
    ...:     try: 
    ...:         assert(len(x.labels) == len(np.unique(img))) 
    ...:     except: 
    ...:         print('{}: number of supplied labels do not match' 
    ...:               ' number of atlas labels'.format(name)) 
    ...:                                                                                      
```
produces
```                         
cort-maxprob-thr50-1mm: atlas labels are not a susquential list of numbers starting at 0
cort-maxprob-thr50-1mm: number of supplied labels do not match number of atlas labels
cort-prob-1mm: number of supplied labels do not match number of atlas labels
cort-prob-2mm: number of supplied labels do not match number of atlas labels
sub-prob-1mm: atlas labels are not a susquential list of numbers starting at 0
sub-prob-1mm: number of supplied labels do not match number of atlas labels
sub-prob-2mm: atlas labels are not a susquential list of numbers starting at 0
sub-prob-2mm: number of supplied labels do not match number of atlas labels
```
This does not seem to be an issue of background (i.e. 0) not being counted in atlas.labels.

I'm not an expert on this atlas. Any suggestions as to why this might be?

**Comment by illdopejake** [comment]
might explain #1871

**Comment by jeromedockes** [comment]
in the thresholded versions of harvard-oxford, some regions may be empty so their label doesn't appear in the image. is that a prolem?

**Comment by illdopejake** [comment]
Good point. That indeed seems to be the problem. I'll get rid of the extra labels then

**Comment by jeromedockes** [comment]
I'm not sure it is a big problem to have empty regions? if we keep them, users will ask why some labels don't appear in the image. if we remove them, users will probably ask why some labels only occur in some versions of the atlas. So I think we mostly need to improve the documentation for this atlas.

**Comment by jeromedockes** [comment]
there will always be the risk that empty regions happen when a user resamples an atlas for example. This is handled correctly by the LabelsMasker, so we need to improve the documentation to explain that the masker's `labels_` attribute should be used to match timeseries with regions, not the atlas' labels

**Comment by illdopejake** [comment]
I looked further into it, and this really only happening for two of the atlases.. the others are just probability images.

Documentation would help but as a naive user I would find it quite counterintuitive to have a label that isn't in the image. Imo, ideally the labels and labels index should be able to be mapped directly to one another, and to the atlas image.

**Comment by bthirion** [comment]
I think that both points of view are good: we should avoid "missing labels" in atlases (yet it is problematic to relabel atlases, e.g. having a nilearn-specific version of Harvard-Oxford), but users should worry about the exact correspondence between labels and extracted features.

**Comment by jeromedockes** [comment]
>  but users should worry about the exact correspondence between labels and extracted features

yes, the documentation should be very clear that this information must be sought in the masker's `labels_` attribute, **not** in the atlas labels

**Comment by illdopejake** [comment]
Fair enough, I'll add a note in the description of Harvard-Oxford, and any other atlases where I find a similar issue.

**Comment by JohannesWiesner** [comment]
I wanted to open an issue related to [my question](https://neurostars.org/t/problems-with-using-niftilabelsmasker/5483) on neurostars.org but I think this is already related to this issue. Please tell me if this should be treated separately, then I can also open up a new issue. As a relatively new user of nilearn I was also very confused to find that the reported number of labels in the atlas-object do not match the number of labels in the data array of the atlas image. The atlas-image for the Destrieux-Altas contains the [reported](https://www.sciencedirect.com/science/article/pii/S1053811910008542?via%3Dihub) 148 regions, but according to the 'labels' recarray object there should be 151 regions (0-region inclusive). 

Here's some code to reproduce this issue:

```
import numpy as np
from nilearn.image import load_img
from nilearn.datasets import fetch_atlas_destrieux_2009

# load destrieux atlas
destrieux_atlas_dict = fetch_atlas_destrieux_2009()

# print information on labels recarray object
# According to this recarray there should be 151 labels (0-label inclusive)
print(destrieux_atlas_dict['labels'])
print('Size of recarray object: {}'.format(destrieux_atlas_dict['labels'].size))

# when looking at the data array of the atlas, yout get the correct 148 regions
# as reported in the paper.
destrieux_atlas = load_img(destrieux_atlas_dict['maps'])
destrieux_atlas_data = destrieux_atlas.get_fdata()
destrieux_atlas_uniques = np.unique(destrieux_atlas_data)
```

**Comment by kchawla-pi** [comment]
@bthirion I am bumping this up, we should make the necessary changes to the documentation now.

**Comment by bthirion** [comment]
@JohannesWiesner  Can you identify the three regions in the recarray that are not found in the image ?
Thx for your help.

**Comment by JohannesWiesner** [comment]
```
# get atlas labels and their region descriptions
destrieux_atlas_dict_labels = [label[0] for label in destrieux_atlas_dict['labels']]
destrieux_atlas_dict_regions = [str(label[1]) for label in destrieux_atlas_dict['labels']]

not_matches = list(set(destrieux_atlas_dict_labels) - set(destrieux_atlas_uniques))
print(not_matches)
```
These two regions appear in the recarray but not in the atlas image:

( 42, b'L Medial_wall') 
(117, b'R Medial_wall')

**Comment by bthirion** [comment]
I feel we should remove these entries from the recarray then.

**Comment by walkerped** [comment]
Did this ever get resolved?  I think I am having a similar issue - when I try to run

```
from nilearn import input_data
from nilearn import datasets

atlas = datasets.fetch_atlas_destrieux_2009(data_dir='/scratch/nilearn')
atlasI = atlas.maps
labelL = atlas.labels
masker = input_data.NiftiLabelsMasker(atlasI,labels=labelL)
masker.fit()
masker.generate_report()
```

```
I get:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/home/scratch/ipykernel_163420/2807402936.py in <module>
      7 masker = input_data.NiftiLabelsMasker(atlasI,labels=labelL)
      8 masker.fit()
----> 9 masker.generate_report()
     10 

/autofs/space/uriline_003/users/projects/OCD_NM/walker_pedersen/scripts/rest/conda/OCDNMrest/lib/python3.9/site-packages/nilearn/input_data/nifti_labels_masker.py in generate_report(self)
    211     def generate_report(self):
    212         from nilearn.reporting.html_report import generate_report
--> 213         return generate_report(self)
    214 
    215     def _reporting(self):

/autofs/space/uriline_003/users/projects/OCD_NM/walker_pedersen/scripts/rest/conda/OCDNMrest/lib/python3.9/site-packages/nilearn/reporting/html_report.py in generate_report(estimator)
    209     else:  # We can create a report
    210         html_template = _get_estimator_template(estimator)
--> 211         overlay, image = _define_overlay(estimator)
    212         parameters = _str_params(estimator.get_params())
    213         docstring = estimator.__doc__

/autofs/space/uriline_003/users/projects/OCD_NM/walker_pedersen/scripts/rest/conda/OCDNMrest/lib/python3.9/site-packages/nilearn/reporting/html_report.py in _define_overlay(estimator)
    151 
    152     """
--> 153     displays = estimator._reporting()
    154 
    155     if len(displays) == 1:  # set overlay to None

/autofs/space/uriline_003/users/projects/OCD_NM/walker_pedersen/scripts/rest/conda/OCDNMrest/lib/python3.9/site-packages/nilearn/input_data/nifti_labels_masker.py in _reporting(self)
    251             if(self.labels is not None
    252                and len(self.labels) != number_of_regions + 1):
--> 253                 raise ValueError(("Mismatch between the number of provided "
    254                                   "labels ({0}) and the number of regions "
    255                                   "in provided label image ({1})").format(

ValueError: Mismatch between the number of provided labels (151) and the number of regions in provided label image (149)
```

**Comment by NicolasGensollen** [comment]
I'm closing this as the mismatch between the length of the list of labels and the number of unique integer values in the image is not a bug (see discussion in #3080).
In order to make things more clear, #3083 will improve the documentation around the atlas fetchers, and we will do some further improvements, listed in #3085.
The report generation issue will be addressed as part of #3085.
